packages/analytics-browser/src/browser-client.ts

```diff
@@ -260,17 +276,40 @@ export class AmplitudeBrowser extends AmplitudeCore implements BrowserClient {
     return super.revenue(revenue, eventOptions);
   }
 
+  private trackCampaignEventIfNeeded(lastEventId?: number, eventsPromise?: AmplitudeReturn<Result>[]) {
+    if (!this.webAttribution || !this.webAttribution.shouldTrackNewCampaign) {
+      return false;
+    }
+    const campaignEvent = this.webAttribution.generateCampaignEvent(lastEventId);
+    if (eventsPromise) {
+      eventsPromise.push(this.track(campaignEvent));
+    } else {
+      this.track(campaignEvent);
+    }
+    this.config.loggerProvider.log('Tracking attribution.');
+    return true;
+  }
+
   async process(event: Event) {
     const currentTime = Date.now();
     const isEventInNewSession = isNewSession(this.config.sessionTimeout, this.config.lastEventTime);
+    const shouldSetSessionIdOnNewCampaign =
+      this.webAttribution && this.webAttribution.shouldSetSessionIdOnNewCampaign();
 
     if (
       event.event_type !== DEFAULT_SESSION_START_EVENT &&
       event.event_type !== DEFAULT_SESSION_END_EVENT &&
-      (!event.session_id || event.session_id === this.getSessionId()) &&
-      isEventInNewSession
+      (!event.session_id || event.session_id === this.getSessionId())
     ) {
-      this.setSessionId(currentTime);
+      if (isEventInNewSession || shouldSetSessionIdOnNewCampaign) {
+        this.setSessionId(currentTime);
+        if (shouldSetSessionIdOnNewCampaign) {
+          this.config.loggerProvider.log('Created a new session for new campaign.');
```

Comment:
How does our logging work? Will this show up always? If yes, should this be loggerProvider.debug?

Reply:
If the logLevel is equal or more than 3, it means if it's either `Verbose, Debug`, we will log this. We do have logger.debug(). But we provide verbose for more possibility. 
[Reference](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/analytics-core/src/logger.ts#L20).
 
```
  log(...args: any[]): void {
    if (this.logLevel < LogLevel.Verbose) {
      return;
    }
    console.log(`${PREFIX}[Log]: ${args.join(' ')}`);
  }
```

Reply:
I kept it as what it was in the web attribution plugin. [Code reference](https://github.com/amplitude/Amplitude-TypeScript/blob/main/packages/plugin-web-attribution-browser/src/web-attribution.ts#L35).